manuals/qcowinfo.1
.Dd June 23, 2026
.Dt QCOWINFO 1
.Os
.Sh NAME
.Nm qcowinfo
.Nd determines information about a QEMU Copy-On-Write (QCOW) image file
.Sh SYNOPSIS
.Nm qcowinfo
.Op Fl hvV
.Ar source
.Sh DESCRIPTION
.Nm qcowinfo
is a utility to determine information about a QEMU Copy-On-Write (QCOW) image \
file
.Pp
.Nm qcowinfo
is part of the
.Nm libqcow
package.
.Nm libqcow
is a library to access the QEMU Copy-On-Write (QCOW) image file format
.Pp
.Ar source
is the source image.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl h
shows this help
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# qcowinfo file.qcow2
qcowinfo 20260623
.sp
QEMU Copy-On-Write (QCOW) image information:
	Format version		: 2
	Media size		: 4.0 GiB (4294967296 bytes)
	Encryption method	: None
	Number of snapshots	: 0
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libqcow/issues
.Sh COPYRIGHT
Copyright (C) 2010-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* pyqcow
  - add function to set the keys
* add backing file support
* add in-image snapshots support

File format:
* reference count support

Tools:
* qcowmount - fuse version
  - fuse multi thread support
* qcowmount - dokan version
  - have it run in background (using CreateProcess)?
  - what about unmount?
  - fix volume size hints
  - create compatible dokan lib

Tests:
* libqcow_file_set_keys
* libqcow_file_set_utf8_password
* libqcow_file_set_utf16_password
* libqcow_get_access_flags functions
* libqcow_get_codepage
* libqcow_set_codepage
* libqcow_check_file_signature and libqcow_check_file_signature_wide
* thread-safety

20130922
* see `git log' for more recent change log
* updated msvscpp files
* removed unused libfvalue

20130916
* worked on libcthreads build support

20130831
* updated dependencies
* bug fixes in libqcow and pyqcow
* added libqcow_io_handle_clear
* worked on tests

20130512
* updated dependencies
* worked on Python bindings
* worked on tests

20130501
* worked on compressed images
* updated msvscpp files
* changed libqcow_file_get_size to libqcow_file_get_media_size
* qcowmount:
  - worked on snapshot support
  - worked on dokan support
* QCOW 2 added support for image with last cluster block smaller
  than cluster block size

20130430
* worked on QCOW 1 support

20130428
* worked on encrypted images
* worked on tests
* worked on QCOW 1 support
* worked on snapshots
* worked on compressed images

20130427
* updated dependencies
* worked on tests

20130422
* updated dependencies
* added python bindings
* added PackageMaker files
* added libqcow_file_get_size

20130330
* updated dependencies
* code clean up

20130310
* 2013 update

20120929
* updated msvscpp files

20120924
* qcowmount added -X option

20120922
* code clean up

20120917
* fix in dpkg files

20120915
* code clean up

20120913
* applied updates
* changes for libcdata update

20120821
* applied updates

20120630
* applied updates

20120522
* fix in dpkg files

20120517
* updated msvscpp files

20120516
* applied updates
* added qcowmount destroy

20120225
* 2012 update
* preparations for handling encrypted and/or compressed cluster blocks

20111012
* update for Cygwin static executable support
* update for libfuse detection

20111009
* updated configure.ac and m4 files
* updated common, libuna, libbfio, libfvalue, libcaes
* implemented info handle
* worked on qcowmount
* updated msvscpp files
* changes to pc and spec file

20110918
* updated configure, array type, list type
* updated libcstring, libsystem, libuna, libbfio, libfdata, libfvalue
* added libcaes

20110304
* updated common, libcstring, libsystem, libuna, libbfio, libfdata, libfvalue
* update for mingw compilation
* added msvscpp files
* qcowinfo: included libintl for Cygwin

20110111
* updated libsystem

20110110
* updated libcstring, libuna, libbfio, libfvalue
* worked on read_buffer
* added seek and read tests

20110109
* updated configure.ac
* synced local libraries
* worked on read_buffer

20110102
* worked on level 1 table support

20101231
* initial version



qcowtools/qcowinfo.c
/*
 * Shows information obtained from a QEMU Copy-On-Write (QCOW) image file.
 *
 * Copyright (C) 2010-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <memory.h>
#include <system_string.h>
#include <types.h>

#include <stdio.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "info_handle.h"
#include "qcowtools_getopt.h"
#include "qcowtools_libcerror.h"
#include "qcowtools_libclocale.h"
#include "qcowtools_libcnotify.h"
#include "qcowtools_libqcow.h"
#include "qcowtools_output.h"
#include "qcowtools_signal.h"
#include "qcowtools_unused.h"

info_handle_t *qcowinfo_info_handle = NULL;
int qcowinfo_abort                  = 0;

/* Signal handler for qcowinfo
 */
void qcowinfo_signal_handler(
      qcowtools_signal_t signal QCOWTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "qcowinfo_signal_handler";

	QCOWTOOLS_UNREFERENCED_PARAMETER( signal )

	qcowinfo_abort = 1;

	if( qcowinfo_info_handle != NULL )
	{
		if( info_handle_signal_abort(
		     qcowinfo_info_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use qcowinfo to determine information about a QEMU Copy-On-Write (QCOW) image file.";

	qcowtools_option_t options[ ] = {
		{ 'h', NULL, "shows this help" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source file" },
	};
	system_character_t options_string[ 32 ];

	libqcow_error_t *error     = NULL;
	system_character_t *source = NULL;
	char *program              = "qcowinfo";
	system_integer_t option    = 0;
	int number_of_options      = (int) ( sizeof( options ) / sizeof( qcowtools_option_t ) );
	int verbose                = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "qcowtools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( qcowtools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	qcowtools_output_version_fprint(
	 stdout,
	 program );

	if( qcowtools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = qcowtools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				qcowtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				qcowtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				qcowtools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		qcowtools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	libqcow_notify_set_stream(
	 stderr,
	 NULL );
	libqcow_notify_set_verbose(
	 verbose );

	if( info_handle_initialize(
	     &qcowinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize info handle.\n" );

		goto on_error;
	}
	if( info_handle_open_input(
	     qcowinfo_info_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source file.\n" );

		goto on_error;
	}
	if( info_handle_file_fprint(
	     qcowinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to print file information.\n" );

		goto on_error;
	}
	if( info_handle_close(
	     qcowinfo_info_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close info handle.\n" );

		goto on_error;
	}
	if( info_handle_free(
	     &qcowinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( qcowinfo_info_handle != NULL )
	{
		info_handle_free(
		 &qcowinfo_info_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

